src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs

```diff
@@ -20,81 +20,199 @@ internal class EventPullerService : BackgroundService
+        public EventGridSubscriptionEventPuller(
+            EventGridTopicSubscription eventGridTopicSubscription,
+            IServiceScopeFactory serviceScopeFactory,
+            ILogger<EventPullerService> logger)
         {
-            try
+            this._serviceScopeFactory = serviceScopeFactory;
+            this._logger = logger;
+            this._eventGridTopicSubscription = eventGridTopicSubscription;
+        }
+
+        public Task StartReceivingEventsAsync(CancellationToken cancellationToken)
+        {
+            // Start the tasks that will process events and handle the completion callback
+            return Task.WhenAll(
+                this.ProcessEvents(cancellationToken),
+                this.AcknowledgeEvents(cancellationToken),
+                this.ReleaseEvents(cancellationToken),
+                this.RejectEvents(cancellationToken));
+        }
+
+        private Task ProcessEvents(CancellationToken cancellationToken)
+        {
+            var parallelOptions = new ParallelOptions { MaxDegreeOfParallelism = this._eventGridTopicSubscription.MaxHandlerDop, CancellationToken = cancellationToken };
+
+            return Parallel.ForEachAsync(this.StreamEventGridEvents(cancellationToken), parallelOptions, async (bundle, ctx) =>
             {
-                var bundles = await eventGridTopicSubscription.Client.ReceiveCloudEventsAsync(eventGridTopicSubscription.TopicName, eventGridTopicSubscription.SubscriptionName, cancellationToken: stoppingToken).ConfigureAwait(false);
-                foreach (var (cloudEvent, lockToken) in bundles)
+                this.SignalHandlerStarting();
+
+                try
+                {
+#pragma warning disable CA2007 // Consider calling ConfigureAwait on the awaited task
+                    await using var scope = this._serviceScopeFactory.CreateAsyncScope();
+#pragma warning restore CA2007 // Consider calling ConfigureAwait on the awaited task
+                    var cloudEventHandler = scope.ServiceProvider.GetRequiredService<ICloudEventHandler>();
+
+                    await this.HandleBundleAsync(cloudEventHandler, bundle.Event, bundle.LockToken, ctx).ConfigureAwait(false);
+                }
+                finally
                 {
-                    await this.HandleEventAsync(cloudEventHandler, eventGridTopicSubscription, cloudEvent, lockToken, stoppingToken).ConfigureAwait(false);
+                    this.SignalHandlerCompleted();
                 }
+            });
+        }
+
+        private async IAsyncEnumerable<EventGridClientAdapter.EventGridClientAdapter.EventBundle> StreamEventGridEvents([EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                await this.WaitForAvailableHandler(cancellationToken).ConfigureAwait(false);
+
+                var availableHandlers = this._eventGridTopicSubscription.MaxHandlerDop - this._handlersInProgress;
```

Comment:
I'm really not sure this is needed

```suggestion
                var availableHandlers = this._eventGridTopicSubscription.MaxHandlerDop - Volatile.Read(ref this._handlersInProgress);
```